src/server/editorServices.ts

```diff
@@ -2626,59 +2629,98 @@ namespace ts.server {
-                                this.refreshScriptInfo(info);
-                            }
+
+                    // Refresh closed script info after an npm install
+                    if (result.refreshScriptInfoRefCount) {
+                        if (dir === fileOrDirectoryPath) {
+                            this.refreshScriptInfosInDirectory(dir);
                         }
-                        // Folder
-                        else if (!hasExtension(fileOrDirectoryPath)) {
-                            this.refreshScriptInfosInDirectory(fileOrDirectoryPath);
+                        else {
+                            const info = this.getScriptInfoForPath(fileOrDirectoryPath);
+                            if (info) {
+                                if (isScriptInfoWatchedFromNodeModules(info)) {
+                                    this.refreshScriptInfo(info);
+                                }
+                            }
+                            // Folder
+                            else if (!hasExtension(fileOrDirectoryPath)) {
+                                this.refreshScriptInfosInDirectory(fileOrDirectoryPath);
+                            }
                         }
                     }
                 },
                 WatchDirectoryFlags.Recursive,
                 this.hostConfiguration.watchOptions,
-                WatchType.NodeModulesForClosedScriptInfo
+                WatchType.NodeModules
             );
-            const result: ScriptInfoInNodeModulesWatcher = {
+            const result: NodeModulesWatcher = {
+                refreshScriptInfoRefCount: affectedModuleSpecifierCacheProject ? 0 : 1,
+                affectedModuleSpecifierCacheProjects: new Set(
+                    affectedModuleSpecifierCacheProject
+                        ? [affectedModuleSpecifierCacheProject.getProjectName()]
+                        : emptyArray),
                 close: () => {
-                    if (result.refCount === 1) {
+                    if (!result.refreshScriptInfoRefCount && !result.affectedModuleSpecifierCacheProjects?.size) {
                         watcher.close();
-                        this.scriptInfoInNodeModulesWatchers.delete(dir);
-                    }
-                    else {
-                        result.refCount--;
+                        this.nodeModulesWatchers.delete(dir);
                     }
                 },
-                refCount: 1
             };
-            this.scriptInfoInNodeModulesWatchers.set(dir, result);
+            this.nodeModulesWatchers.set(dir, result);
             return result;
         }
 
+        /*@internal*/
+        watchPackageJsonsInNodeModules(dir: Path, project: Project): FileWatcher {
+            const existing = this.nodeModulesWatchers.get(dir);
+            if (existing) {
+                (existing.affectedModuleSpecifierCacheProjects ||= new Set()).add(project.getProjectName());

```

Comment:
isnt this incorrect, you would want to return the watcher.close function that does what it is doing below instead of returning the actual watcher created by nodeModules watcher?

Reply:
Yikes, good catch; totally missed this in a later refactor.